File: src/org/cerion/symcalc/expression/function/integer/PowerMod.java
package org.cerion.symcalc.expression.function.integer;

import org.cerion.symcalc.exception.ValidationException;
import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.number.IntegerNum;

public class PowerMod extends FunctionExpr {

	public PowerMod(Expr ...e) {
		super(FunctionType.POWERMOD, e);
	}
	
	@Override
	public Expr evaluate() {
		if(get(0).isInteger() && get(1).isInteger() && get(2).isInteger())
		{
			IntegerNum a = (IntegerNum)get(0);
			IntegerNum b = (IntegerNum)get(1);
			IntegerNum c = (IntegerNum)get(2);
			return a.powerMod(b, c);
		}

		return this;
	}

	@Override
	public void validate() throws ValidationException {
		validateParameterCount(3);
		validateParameterType(0, ExprType.NUMBER);
		validateParameterType(1, ExprType.NUMBER);
		validateParameterType(2, ExprType.NUMBER);
	}
}

File: src/main/kotlin/function/integer/PowerMod.kt
package org.cerion.symcalc.function.integer

import org.cerion.symcalc.exception.ValidationException
import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.function.FunctionExpr
import org.cerion.symcalc.number.NumberType

class PowerMod(vararg e: Expr) : FunctionExpr(*e) {

    public override fun evaluate(): Expr {
        val a = get(0).asInteger()
        val b = get(1).asInteger()
        val c = get(2).asInteger()
        return a.powerMod(b, c)
    }

    @Throws(ValidationException::class)
    override fun validate() {
        validateNumberType(0, NumberType.INTEGER)
        validateNumberType(1, NumberType.INTEGER)
        validateNumberType(2, NumberType.INTEGER)
    }
}
